### PR TITLE
Add support for EDAlias as a case in SwitchProducer

### PR DESCRIPTION
**Pull request by makortel**
In a heterogeneous workflow it may happen that the (legacy) CPU and non-CPU modules are naturally defined so differently that the final legacy CPU products are produced in a very asymmetric way (e.g.
* CPU: producer `A` produces `a1`, `a2`, producer `B` produces `b1
* non-CPU: producer `D` produces `a1`, producer `C` produces `a2` and `b1`

). In order to be able to handle such cases flexibly and efficiently, this PR extends the SwitchProducer mechanism to allow EDAliases as cases (in addition to EDProducers). This required first some fixes to EDAliases:
* Throw if an EDAlias defines two instances pointing to the same real product
   * Currently the OutputModule will complain if an alias and the alias-for branches are attempted to be kept. This check does not work with SwitchProducer, whose output branches are not exactly alias branches (even if they're very close to). The premise is that it does not make sense to have both the alias and alias-for branches.
* If a job has an unscheduled EDProducer that registers a callback for new product registration such that the callback calls `produces()`, and that callback gets triggered by EDAlias (output) BranchDescriptions, the EDProducer BranchDescriptions have an incorrect on-demand status leading to a wrong ProductResolver and eventually a job to hang because of starvation in the TBB scheduler. This situation happens because the on-demand status for EDProducer BranchDescriptions is set before the EDAliases are processed in the Schedule. The fix is to move the setting of the on-demand status of BranchDescriptions to be done after the EDAlias output BranchDescriptions have been created.
  * The setting is also a bit more efficient now that it is done once instead of once per stream
* Made the python `EDAlias` cloneable and to inherit from `_Parameterizable`
  * Latter reduces copy-paste and enables `Modifier.toReplaceWith()`

Adding the support for EDAlias cases in SwitchProducer required the following changes
* SwitchProducer ProductResolvers can be set only after all EDAlias ProductResolvers have been set
* Added a new (intermediate) base class `DataManagingOrAliasProductResolver`, from which the `DataManaging`, `Alias`, and `SwitchBase` ProductResolvers inherit from, in order to be able use both `DataManaging` and `Alias` ProductResolvers as the `realProduct` in `SwitchBaseProductResolver`
* Added a special treatment for SwitchProducer in `ProductSelector` to detect the case when one attempts to keep in OutputModule both the real product, and the EDAlias-within-SwitchProducer. Similar check is already done for regular EDAliases (as mentioned above)
  * The same check does not work for SwitchProducers because there the `BranchID` of the SwitchProducer and the case product/alias are different, while in case of EDAlias the `BranchID` of the alias and the alias-for product as the same.
* In python `Process`, separate `SwitchProducers` from the normal `EDProducers`. This change is needed in order to append elements to both `@all_modules` and `@all_aliases` `vstring`s when filling the ProcessDescription. This way the case-EDAlias can away from the `Process.aliases_()` (as the case-EDProducers are not in `Process.producers_()`).
* As an additional sanity check, mark all case BranchDescriptions as transient in order to ignore them in OutputModule keep statements like `keep *_foo*_*_*`.
  * I also tried to "implicitly drop" all branches with '@' in `ProductSelector`, but that lead to SwitchProducer branches being dropped on inputs as they would be interpreted as branches depending on a dropped branch.

This PR also adds a generic mechanism to allow customized configuration entities to pass import statements to configuration dump. Related, currently the case-producer fragments end up in the configuration dump while they should not, this PR fixes that (as a byproduct).

Tested in CMSSW_10_5_X_2019-02-08-1100, no changes expected in current workflows.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: FWCore/Framework/src/Schedule.cc [hunk 1]
```diff
@@ -620,7 +644,38 @@ namespace edm {
     std::map<std::string, std::vector<std::pair<std::string, int> > > outputModulePathPositions;
     reduceParameterSet(proc_pset, tns.getEndPaths(), modulesInConfig, usedModuleLabels,
                        outputModulePathPositions);
-    processEDAliases(proc_pset, processConfiguration->processName(), preg);
+    const bool hasAliases = processEDAliases(proc_pset, processConfiguration->processName(), preg);
+
+    // At this point all BranchDescriptions are created. Mark now the
+    // ones of unscheduled workers to be on-demand.
+    // TODO: what to do for alias branches???
```
Comment (Dr15Jones) [inline FWCore/Framework/src/Schedule.cc]: What is the plan to resolve this TODO?

File: FWCore/Integration/test/run_TestEDAlias.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+test=testEDAlias
```
Comment (Dr15Jones) [inline FWCore/Integration/test/run_TestEDAlias.sh]: How is this test run? I didn't notice anything in BuildFile.xml.

**Review by Dr15Jones — CHANGES_REQUESTED** [review]

File: FWCore/Framework/src/Schedule.cc (hunk 1 above)
Comment (makortel) [reply FWCore/Framework/src/Schedule.cc]: Thanks for catching, the comment itself is a leftover. The alias branches are already dealt with below on lines 667-676.

File: FWCore/Integration/test/run_TestEDAlias.sh (hunk 2 above)
Comment (makortel) [reply FWCore/Integration/test/run_TestEDAlias.sh]: Sigh, I forgot to add it again to BuildFile.xml, thanks for catching.

**Comment by makortel** [comment]
Addressed @Dr15Jones' comments, fixed the compilation failure, and added a check for an additional corner case.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
At least the `testGroupSelection` works fine locally on CMSSW_10_5_X_2019-02-08-1100.

**Comment by makortel** [comment]
@cmsbuild, please test

Let's see if a new IB would make any difference.

**Comment by fwyzard** [comment]
as a general comment, it would be very nice to have this in pre2 ...

**Comment by makortel** [comment]
Only one unit test failure?
```
%MSG-w SiStripCabling:  SiStripO2OBadStrip:siStripO2O 13-Feb-2019 19:44:03 CET  Run: 1 Event: 1
[FedChannelConnection::apvPairNumber] Unexpected number of APV pairs: 65535
%MSG


[2019-02-13 19:44:03,856] INFO: @@@CMSSW job return code = -11@@@
Traceback (most recent call last):
  File "/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-13-1100/bin/slc7_amd64_gcc700/SiStripDAQPopCon.py", line 181, in <module>
    main()
  File "/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-13-1100/bin/slc7_amd64_gcc700/SiStripDAQPopCon.py", line 174, in main
    runjob(args)
  File "/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-13-1100/bin/slc7_amd64_gcc700/SiStripDAQPopCon.py", line 89, in runjob
    raise RuntimeError('O2O job FAILED!')
RuntimeError: O2O job FAILED!
Failure running SiStripDAQPopCon.py: status 1
```

The error in 250202.181 step2 is essentially a file read error.

**Comment by makortel** [comment]
@cmsbuild, please test

I want to see how well the behaviour reproduces...

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
Added two things for configuration
* unit test that the case-producer configuration fragments (with `@` in the module label) do not end up in the configuration dump
* a generic mechanism to allow customized configuration entities to pass import statements to configuration dump
  * this is a bit of feature creep for this PR, but technically the test of this feature conflicts with the previous test addition, and overall this PR is needed to remove the case-producer fragments from the dump

**Comment by makortel** [comment]
The error in runtestPhysicsToolsPatAlgos is again one of the weird ones
```
#4  <signal handler called>
#5  0x00007f03cbcb782c in std::_Rb_tree<edm::TypeIDBase, std::pair<edm::TypeIDBase const, std::shared_ptr<edm::serviceregistry::ServiceWrapperBase> >, std::_Select1st<std::pair<edm::TypeIDBase const, std::shared_ptr<edm::serviceregistry::ServiceWrapperBase> > >, std::less<edm::TypeIDBase>, std::allocator<std::pair<edm::TypeIDBase const, std::shared_ptr<edm::serviceregistry::ServiceWrapperBase> > > >::find(edm::TypeIDBase const&) () from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-14-1100/lib/slc7_amd64_gcc700/libFWCoreFramework.so
#6  0x00007f03cbf872e4 in edm::serviceregistry::ServicesManager::~ServicesManager() () from /cvmfs/cms-ib.cern.ch/nweek-02563/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_X_2019-02-13-2300/lib/slc7_amd64_gcc700/libFWCoreServiceRegistry.so
#7  0x00007f03cbf87421 in edm::serviceregistry::ServicesManager::~ServicesManager() () from /cvmfs/cms-ib.cern.ch/nweek-02563/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_X_2019-02-13-2300/lib/slc7_amd64_gcc700/libFWCoreServiceRegistry.so
#8  0x0000000000413786 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() ()
#9  0x00007f03cbc5db8a in edm::EventProcessor::~EventProcessor() () from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-14-1100/lib/slc7_amd64_gcc700/libFWCoreFramework.so
#10 0x000000000040e6e3 in (anonymous namespace)::EventProcessorWithSentry::~EventProcessorWithSentry() ()
#11 0x000000000040e17f in main ()

Current Modules:

Module: none (crashed)
```
happening at the very end of the job.

**Comment by Dr15Jones** [comment]
@makortel please document in the comments what tests you've run to try to find the cause of the unit test failures (e.g. what tools you ran). Once documented, I will sign off on the pull request assuming the failures are unrelated.

**Comment by makortel** [comment]
I ran all the failed tests in https://github.com/cms-sw/cmssw/pull/25907#issuecomment-462590189 in a local area in my development IB and in the IB used in the PR tests, all of those tests succeeded.

In #25914 I ran runtestRecoEgammaPhotonIdentification with valgrind, but got nothing relevant.

I'm still planning to play with ASAN for some of the tests in this PR. I'll anyway document all findings in this PR.

**Comment by Dr15Jones** [comment]
+1
The unit test failures are also appearing in other pull requests as well.

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
I tested the following tests in CMSSW_10_5_ASAN_X_2019-02-08-2300 (didn't have time to run all tests until this ASAN IB expired)

```
---> test testTauEmbeddingProducers had ERRORS
---> test runtestTqafTopEventProducers had ERRORS
---> test runtestPhysicsToolsNanoAOD had ERRORS
---> test runtestTqafTopKinFitter had ERRORS
---> test runtestRecoEgammaPhotonIdentification had ERRORS
---> test runtestPhysicsToolsPatAlgos had ERRORS
---> test runtestRecoEgammaElectronIdentification had ERRORS
---> test testPhase2PixelNtuple had ERRORS
---> test runtestTqafExamples had ERRORS
```

**Comment by Dr15Jones** [comment]
@civanch @mdhildreth please review. We believe the test failures are from problems in the pull request testing system when too many tests have to be run and are NOT from this pull request.

**Comment by fabiocos** [comment]
please test

**Comment by civanch** [comment]
@makortel , for me it is a bit strange that this PR trigger problems of these tests. May it be that the modification makes the tests to be sensitive and they cannot work in parallel at the same node?

**Comment by makortel** [comment]
@civanch I fully agree that the unit test failures are strange. But the failures are somewhat random (e.g. in https://github.com/cms-sw/cmssw/pull/25907#issuecomment-463367248 only `SiStripDAQ_O2O_test` failed, and 250202.181 because of file open errors), occur also in other PRs, and are not reproducible in local tests. 

Also @smuzaffar commented in https://github.com/cms-sw/cmssw/pull/25917#issuecomment-464771140
> We have noticed that when PR include too many packages (e.g. in this case over 700) then unit tests misbehave and are mostly killed by `timeout`

**Comment by fabiocos** [comment]
I confirm that in my local test build I manage to execute successfully the failing tests. Let's try again...

**Comment by boudoul** [comment]
Dear all, please note that the PR #25736 is also experiencing strange and random unit test failures ....

**Comment by fabiocos** [comment]
code-checks

**Comment by fabiocos** [comment]
@civanch @mdhildreth as far as I can see, the change under the simulation umbrella in Mixing/Base is just a direct consequence of the other updates (the method used has been replace by another one). The short matrix tests involving PU do not show any difference, and the random failures in the unit tests are not explicitly related to the SIM part, and looks most likely an issue with the test ssyetm itself (under investigation).

In agreement with @smuzaffar I will move forward with the merge of this PR, so as we may verify in an IB whether we observe similar problems (none so far).

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

to verify in an IB environment whether the observed random crashes appear as well, or whether the problem is related to the PR test environment

**Comment by makortel** [comment]
Seems that all unit tests indeed succeeded in the IB.

**Comment by fabiocos** [comment]
@makortel indeed this points towards an issue in the test system...
